### PR TITLE
Fixed absolute url generation for query strings and hash urls

### LINKED ISSUE #23260: Absolute url twig extension not working with query strings only
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? |no
| BC Break report? | yes?
| RFC?             |no
| Symfony version  | > 2.8 <= 3.3?

`absolute_url` twig extension not working with query strings `?test=1` only.

```twig
{{ absolute_url('?foo=bar') }}
```

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q             | A
| ------------- | ---
| Branch?       | 2.7, ...
| Bug fix?      | yes
| New feature?  |no 
| BC breaks?    | yes? absolute_url will change its output but the old was incorrect
| Deprecations? |no 
| Tests pass?   | yes?
| Fixed tickets | fixes #23260
| License       | MIT

Fixed absolute url generation for query strings

### CODE REVIEW

**Comment by stof** [comment]
would be great to check if the Link URL resolution in DomCrawler suffers from the same issue

**Comment by alexander-schranz** [comment]
@stoff Looks good there but seems we need to handle `#` also correctly in the absolute_url. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DomCrawler/AbstractUriElement.php#L100

**Comment by alexander-schranz** [comment]
will change the base to 2.7 think thats the oldest supported version?

**Comment by alexander-schranz** [comment]
I'm not sure why php7 keeps failing on travis

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 1]
```diff
@@ -70,6 +70,15 @@ public function generateAbsoluteUrl($path)
                     $port = ':'.$this->requestContext->getHttpsPort();
                 }
 
+                if ('#' === $path[0]) {
+                    $queryString = $this->requestContext->getQueryString();
+                    $path = $this->requestContext->getPathInfo().($queryString ? '?'.$queryString : '').$path;
+                }
+
+                if ('?' === $path[0]) {
```
Comment (dmaicher) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: We can use `elseif` here and also below?

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 2]
```diff
@@ -80,6 +89,14 @@ public function generateAbsoluteUrl($path)
             return $path;
         }
 
+        if ('#' === $path[0]) {
+            $path = $request->getRequestUri().$path;
+        }
+
+        if ('?' === $path[0]) {
```
Comment (dmaicher) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: also `elseif`?

File: src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php [hunk 3]
```diff
@@ -41,6 +41,14 @@ public function getGenerateAbsoluteUrlData()
...
             array('//example.com/baz', '//example.com/baz', '/'),
+
+            array('http://localhost/foo/bar?baz', '?baz', '/foo/bar'),
+            array('http://localhost/foo/bar?baz=1', '?baz=1', '/foo/bar?foo=1'),
+            array('http://localhost/foo/baz?baz=1', 'baz?baz=1', '/foo/bar?foo=1'),
+
+            array('http://localhost/foo/bar#baz', '#baz', '/foo/bar'),
+            array('http://localhost/foo/bar?baz=1#baz', '?baz=1#baz', '/foo/bar?foo=1'),
+            array('http://localhost/foo/baz?baz=1#baz', 'baz?baz=1#baz', '/foo/bar?foo=1'),
         );
```
Comment (dmaicher) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php]: can you add this test case?

```php
array('http://localhost/foo/bar?0#foo', '#foo', '/foo/bar?0'),
```

I think your current solution will strip out the existing `?0` query string. I know its an edge case but still :smile:

Comment (alexander-schranz) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php]: test added seems not failing as a `0` as string is true.

Comment (dmaicher) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php]:
```sh
php -r "var_dump((bool)'0');"
```
==> `bool(false)`

Comment (alexander-schranz) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php]: @dmaicher can you tell me whats wrong with the test as it should fail.

Comment (dmaicher) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpFoundationExtensionTest.php]: Ah this is only executed when no current request is given (CLI) and the request context is used instead :wink: 

So your changes from L 73-79 are currently not covered by tests. You need to add some more test cases into the `getGenerateAbsoluteUrlRequestContextData` provider :wink:

**Comment by Tobion** [comment]
👍

**Review by Tobion — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexander-schranz.

**Pull request closed without merging**